core/src/main/java/org/apache/iceberg/io/BaseDeltaWriter.java

```diff
@@ -0,0 +1,248 @@
+                         RollingContentFileWriter<DeleteFile, PositionDelete<T>> posDeleteWriter,
+                         RollingContentFileWriter<DeleteFile, T> equalityDeleteWriter,
+                         Schema tableSchema,
+                         List<Integer> equalityFieldIds,
+                         Function<T, StructLike> structLikeFun) {
+
+    Preconditions.checkNotNull(dataWriter, "Data writer should always not be null.");
+
+    if (posDeleteWriter == null) {
+      // Only accept INSERT records.
+      Preconditions.checkArgument(equalityDeleteWriter == null);
+    }
+
+    if (posDeleteWriter != null && equalityDeleteWriter == null) {
+      // Only accept INSERT records and position deletion.
+      Preconditions.checkArgument(tableSchema == null);
+      Preconditions.checkArgument(equalityFieldIds == null);
+    }
+
+    if (equalityDeleteWriter != null) {
+      // Accept insert records, position deletion, equality deletions.
+      Preconditions.checkNotNull(posDeleteWriter,
+          "Position delete writer shouldn't be null when writing equality deletions.");
+      Preconditions.checkNotNull(tableSchema, "Iceberg table schema shouldn't be null");
+      Preconditions.checkNotNull(equalityFieldIds, "Equality field ids shouldn't be null");
+      Preconditions.checkNotNull(structLikeFun, "StructLike function shouldn't be null");
+
+      Schema deleteSchema = TypeUtil.select(tableSchema, Sets.newHashSet(equalityFieldIds));
+      this.insertedRowMap = StructLikeMap.create(deleteSchema.asStruct());
+      this.structLikeFun = structLikeFun;
+    } else {
+      this.insertedRowMap = null;
+      this.structLikeFun = null;
+    }
+
+    this.dataWriter = dataWriter;
+    this.equalityDeleteWriter = equalityDeleteWriter;
+    this.posDeleteWriter = posDeleteWriter;
+  }
+
+  @Override
+  public void writeRow(T row) throws IOException {
+    if (enableEqualityDelete()) {
+      FilePos filePos = FilePos.create(dataWriter.currentPath(), dataWriter.currentPos());
+      insertedRowMap.compute(structLikeFun.apply(row), (k, v) -> {
+        if (v == null) {
+          return Lists.newArrayList(filePos);
+        } else {
+          v.add(filePos);
+          return v;
+        }
+      });
+    }
+
+    dataWriter.write(row);
+  }
+
+  @Override
+  public void writeEqualityDelete(T equalityDelete) throws IOException {
+    if (!enableEqualityDelete()) {
+      throw new UnsupportedOperationException("Could not accept equality deletion.");
+    }
+
+    List<FilePos> existing = insertedRowMap.get(structLikeFun.apply(equalityDelete));
+
+    if (existing == null) {
+      // Delete the row which have been written by other completed delta writer.
+      equalityDeleteWriter.write(equalityDelete);
+    } else {
+      // Delete the rows which was written in current delta writer.
+      for (FilePos filePos : existing) {
+        posDeleteWriter.write(positionDelete.set(filePos.path(), filePos.pos(), null));
```

Comment:
I think this strategy will violate the ordering requirements of position delete files.

Position delete files need to be ordered by file and position. Imagine 3 replacements for the same row key. The first creates insert 0, the second insert 1, and the third insert 2. The first adds 0 existing, the second adds 1, and the third adds 2. And the first time through this loop deletes position 0, the second deletes positions 0 and 1, and the third deletes positions 0, 1, and 2. The result is deleting positions 0, 0, 1, 0, 1, 2, which violates the order requirement.

For a given inserted position, we need to track that it is deleted and write all of the deletes at one time so that they are all in order.

I think this should change from using `StructLikeMap<List<FilePos>>` to `StructLikeMap<FilePos>`. Each time the row is deleted, get the current `FilePos` and add it to a `deletePositions` list. Then set the new row's `FilePos` in the map so that it is the one deleted the next time.

Reply:
Great point !